### ISSUE TITLE
Improve the meilisearch integration tests

### ISSUE THREAD

**Issue by irevoire** [meilisearch/meilisearch#4840]
Hello, fellow contributors! Thanks for coming here 🌟 

In https://github.com/meilisearch/meilisearch/pull/4808, I started some work to make the test faster but didn't apply the techniques implemented to every test since there are so many of them, and that's where you can help us.

There are two tasks to do that I’ll highlight right below.

### Patching the `.wait_task(:task_id)`

A common pattern you'll see in the integration tests is that we create a task and then immediately wait for it by writing `.wait_task(0).await;` or any other number.
Here's an example: https://github.com/meilisearch/meilisearch/blob/abe128476fca60d683725f2598a365027ef5fbec/meilisearch/tests/tasks/mod.rs#L111-L134

This is really easy to get wrong and caused us a lot of flaky tests in the past; that's why we would like to get rid of all these calls and replace them with the actual task ID returned by Meilisearch.

For example, instead of doing:
```rust
    let index = server.index("test");
    index.create(None).await;
    index.wait_task(0).await;
```

We should write:
```rust
    let index = server.index("test");
    let (task, _code) = index.create(None).await;
    index.wait_task(task.uid()).await;
```

Also, did you notice how we're waiting for the task to finish, but then we don't check if it succeeded?
When the task was supposed to succeed in the tests, you should also call `.succeeded()` after waiting for the task:

```rust
    let index = server.index("test");
    let (task, _code) = index.create(None).await;
    index.wait_task(task.uid()).await.succeeded();
```

### Using a shared server when possible

A lot of tests are creating a new Meilisearch instance when they could instead share one with all the other tests.
The rules to know if a test can use a shared server are:
- It should use the default configuration; the server is initialized with `let server = Server::new().await;`
- It must **not** modify the global state of the index:
  - Don't add an API-key
  - Don't modify the features
  - Don't access indexes that are not yours
- ❗ An exception, though, is if the request is expected to fail in the test. For example, if the test try to delete an arbitrary index and then checks for its failure, in this case, we can implement a new method on the shared index, send the task, and ensure it fails; see this method for an example: https://github.com/meilisearch/meilisearch/blob/c45706936732aef16d5a07545e47534d59901a91/meilisearch/tests/common/server.rs#L218-L225

https://github.com/meilisearch/meilisearch/blob/abe128476fca60d683725f2598a365027ef5fbec/meilisearch/tests/search/geo.rs#L41-L63

This test, for example, only creates an index and modifies its content without altering any of the global state of the index.
We can update the server initialization from:
```rust
    let server = Server::new().await;
    let index = server.index("test");
```

To:

```rust
    let server = Server::new_shared().await;
    let index = server.unique_index(); // we cannot hardcode the name of an index; otherwise, we might have conflicts between tests
```

### Using a shared index when possible

The latest kind of big optimization possible is to reduce the number of times we index the same documents over and over again.
For example, if you look at this file: https://github.com/meilisearch/meilisearch/blob/abe128476fca60d683725f2598a365027ef5fbec/meilisearch/tests/search/geo.rs
You'll notice a common pattern: All the tests start by doing these operations and then do a single search request:
```rust
    let server = Server::new().await;
    let index = server.index("test");

    let documents = DOCUMENTS.clone();
    index.update_settings_filterable_attributes(json!(["_geo"])).await;
    index.update_settings_sortable_attributes(json!(["_geo"])).await;
    index.add_documents(documents, None).await;
    index.wait_task(2).await;
```

What we should do here is create a shared index between all these tests that index the documents only once and can then be used by everyone. Here's an example of how to do it: https://github.com/meilisearch/meilisearch/blob/c45706936732aef16d5a07545e47534d59901a91/meilisearch/tests/common/mod.rs#L200-L216

In this case, instead of creating a new set of `DOCUMENTS` and a new kind of `shared_index_with_geo` we could simply update the original `DOCUMENTS` set defined [here](https://github.com/meilisearch/meilisearch/blob/c45706936732aef16d5a07545e47534d59901a91/meilisearch/tests/common/mod.rs#L165-L198) and update then use the shared index function to get an index.

In the end, the test should look like this:

```rust
let index = shared_index_with_documents().await;

// The search requests
```

Once again, this optimization can only be applied to tests that don't modify the content of the index:
- Can't insert or update documents
- Can't delete documents
- Can't update the settings
- Can almost exclusively run searches, get documents, and stuff like that
- ❗ An exception, though, is if the request is expected to fail in the test. For example, if the test adds a wrong setting and then checks for its failure, in this case, we can implement a new method on the shared index that runs send the task and then ensures it fails; see this method for an example: https://github.com/meilisearch/meilisearch/blob/c45706936732aef16d5a07545e47534d59901a91/meilisearch/tests/common/index.rs#L259-L274

### The file to updates

Here's an almost exhaustive list of file that need to be updated:

- [ ] meilisearch/tests/auth/tenant_token.rs
- [ ] meilisearch/tests/auth/tenant_token_multi_search.rs
- [ ] meilisearch/tests/documents/add_documents.rs
- [ ] meilisearch/tests/documents/delete_documents.rs
- [ ] meilisearch/tests/documents/errors.rs
- [ ] meilisearch/tests/documents/get_documents.rs
- [ ] meilisearch/tests/documents/update_documents.rs
- [ ] meilisearch/tests/index/create_index.rs
- [ ] meilisearch/tests/index/delete_index.rs
- [ ] meilisearch/tests/index/get_index.rs
- [ ] meilisearch/tests/index/stats.rs
- [ ] meilisearch/tests/index/update_index.rs
- [ ] meilisearch/tests/search/errors.rs
- [ ] meilisearch/tests/search/facet_search.rs
- [ ] meilisearch/tests/search/formatted.rs
- [ ] meilisearch/tests/search/locales.rs
- [ ] meilisearch/tests/search/matching_strategy.rs
- [ ] meilisearch/tests/search/mod.rs
- [ ] meilisearch/tests/search/multi.rs
- [ ] meilisearch/tests/search/pagination.rs
- [ ] meilisearch/tests/search/restrict_searchable.rs
- [ ] meilisearch/tests/settings/distinct.rs
- [ ] meilisearch/tests/settings/get_settings.rs
- [ ] meilisearch/tests/settings/proximity_settings.rs
- [ ] meilisearch/tests/settings/tokenizer_customization.rs
- [ ] meilisearch/tests/stats/mod.rs
- [ ] meilisearch/tests/tasks/mod.rs

### What to do if I’m struggling

Please ping me and ask me questions on this issue; this will help others as well.

**Comment by chenlong1993** [comment]
I want to give it a try

**Comment by irevoire** [comment]
Thanks, @chenlong1993, that’s great.
To avoid getting overwhelmed by the quantity of work, I think it’s best if you could split your work into one PR per file.

**Comment by sunnynagavo** [comment]
Hi @irevoire , I have submitted PR for fixing one file from your above list. could you please take a look and provide me with feedback. 

https://github.com/meilisearch/meilisearch/pull/4865

**Comment by DerTimonius** [comment]
working on `meilisearch/tests/index/create_index.rs`

**Comment by Gnosnay** [comment]
hi there, may i know whether this issue is fixed totally?

**Comment by irevoire** [comment]
Yes it's up to date 👍

**Comment by irevoire** [comment]
Hey, yes, definitely there are a ton of improvements we could try to do on each test. Here, this issue is more focused on the time we spent running all the tests at the same time. You can see it by running the test suite under a profiler.
A lot of time is spent creating and removing threads every time we create and drop a meilisearch client.
In this issue, the idea is to use a shared client for as many tests as possible so we don't create new threads.

**Comment by martin-g** [comment]
@irevoire I think `meilisearch/tests/search/locales.rs` should be removed from the list above. All its tests modify the settings, e.g. 
```rust
index
        .update_settings(json!({
            "searchableAttributes": ["product"],
            "localizedAttributes": [
                // force swedish
                {"attributePatterns": ["product"], "locales": ["deu"]}
            ]
        }))
        .await;
```

**Comment by irevoire** [comment]
Hey, but the settings are not global, they are specific to each index, which means we can still share a single `Server`, right?

**Comment by martin-g** [comment]
@irevoire All the listed tests but the Auth tenant ones are addressed!

About the Auth tenant ones:  `server.use_api_key(...)` makes it hard to reuse a shared server.
We can add `server.use_api_key("MASTER_KEY");` to `new_shared_with_admin_key()` but https://github.com/meilisearch/meilisearch/blob/abb399b802f4ed54d5849f4deac10c2c71a12477/crates/meilisearch/tests/auth/tenant_token_multi_search.rs#L298 sets web token ones and this would cause problems.
Ideas ?

**Comment by irevoire** [comment]
>  sets web token ones and this would cause problems.

Yeah, I feel like you're right, and it doesn't really make sense to try to test something that is always global anyway. Most of the tests won't work nicely.
I think I should remove them from the list. What do you think?

**Comment by martin-g** [comment]
I think the same!

I have some more ideas for minor improvements. I'll send PRs soon!

**Comment by martin-g** [comment]
I have opened:
* https://github.com/meilisearch/meilisearch/pull/5697
* https://github.com/meilisearch/meilisearch/pull/5698
* https://github.com/meilisearch/meilisearch/pull/5699
* https://github.com/meilisearch/meilisearch/pull/5700

Once merged I will open the last one that removes Index::wait_task() completely and updates the rest of the files.

**Comment by martin-g** [comment]
> Once merged I will open the last one that removes Index::wait_task() completely and updates the rest of the files.

https://github.com/meilisearch/meilisearch/pull/5703
